File: runtime/java/units/wasi/path_link.java
// requires: memory/read_string, wasi/resolve_path, wasi/errno_fs
// Create a hard link. Both endpoints are confined to their dir fds and
// resolved NOFOLLOW, so the link is made to the source name itself (a symlink
// source is hard-linked as the symlink, not its target). Following the source
// symlink (LOOKUPFLAGS_SYMLINK_FOLLOW) is rejected; hard-linking a directory is
// EPERM; a trailing slash on the destination is ENOENT.
//
// macOS link(2) follows a symlink source (unlike Linux and unlike the
// AT_SYMLINK_NOFOLLOW linkat the suite expects), and Files.createLink cannot
// express NOFOLLOW, so createLink on a symlink source follows the (dangling)
// link and throws. Emulate a NOFOLLOW hard-link-to-a-symlink by recreating the
// symlink at the destination (mirrors the Go backend's fallback).
int wasi_path_link(int oldFd, int oldFlags, int oldPtr, int oldLen, int newFd, int newPtr,
                   int newLen) {
    if ((oldFlags & 0x1) != 0) { // lookupflags::SYMLINK_FOLLOW
        return WASI_INVAL;
    }
    String oldRel = new String(
        memory.read_string(Integer.toUnsignedLong(oldPtr), Integer.toUnsignedLong(oldLen)),
        java.nio.charset.StandardCharsets.UTF_8);
    String newRel = new String(
        memory.read_string(Integer.toUnsignedLong(newPtr), Integer.toUnsignedLong(newLen)),
        java.nio.charset.StandardCharsets.UTF_8);
    Resolved oldR = resolve_path(oldFd, oldRel, false);
    if (oldR.errno != WASI_OK) {
        return oldR.errno;
    }
    Resolved newR = resolve_path(newFd, newRel, false);
    if (newR.errno != WASI_OK) {
        return newR.errno;
    }
    if (newRel.endsWith("/")) {
        return WASI_NOENT;
    }
    java.nio.file.Path oldP = java.nio.file.Paths.get(oldR.path);
    java.nio.file.Path newP = java.nio.file.Paths.get(newR.path);
    if (java.nio.file.Files.isDirectory(oldP, java.nio.file.LinkOption.NOFOLLOW_LINKS)) {
        return WASI_PERM;
    }
    try {
        java.nio.file.Files.createLink(newP, oldP);
    } catch (java.io.IOException ex) {
        if (java.nio.file.Files.isSymbolicLink(oldP)) {
            try {
                java.nio.file.Files.createSymbolicLink(newP, java.nio.file.Files.readSymbolicLink(oldP));
                return WASI_OK;
            } catch (java.io.IOException se) {
                return fs_errno(se);
            }
        }
        return fs_errno(ex);
    }
    return WASI_OK;
}

File: runtime/ruby/units/wasi/path_link.rb
# requires: memory/read_string, wasi/resolve_path, wasi/errno_fs
def wasi_path_link(old_dirfd, old_flags, old_ptr, old_len, new_dirfd, new_ptr, new_len)
  # path_link never dereferences the source symlink; a SYMLINK_FOLLOW
  # request is rejected outright.
  return ERRNO_INVAL if old_flags & 0x1 != 0
  old_rel = @memory.read_string(old_ptr, old_len)
  new_rel = @memory.read_string(new_ptr, new_len)
  # A trailing slash on the new name would name a directory, which a
  # hardlink can never be.
  return ERRNO_NOENT if new_rel.end_with?("/")
  old_host, err = resolve_path(old_dirfd, old_rel, follow_last: false)
  return err if err
  new_host, err = resolve_path(new_dirfd, new_rel, follow_last: false)
  return err if err
  errno = linkat_nofollow(old_host, new_host)
  raise SystemCallError.new("path_link", errno) unless errno.zero?
  ERRNO_SUCCESS
rescue SystemCallError => e
  fs_errno(e)
end

# link(2) follows symlinks on macOS/BSD, but WASI's path_link (without
# SYMLINK_FOLLOW) must hardlink the link itself. linkat(..., 0) is nofollow
# on every POSIX.1-2008 platform, so bind it via Fiddle; returns 0 on
# success or the errno on failure.
def linkat_nofollow(old_host, new_host)
  require "fiddle"
  @linkat ||= Fiddle::Function.new(
    Fiddle::Handle::DEFAULT["linkat"],
    [Fiddle::TYPE_INT, Fiddle::TYPE_VOIDP, Fiddle::TYPE_INT, Fiddle::TYPE_VOIDP, Fiddle::TYPE_INT],
    Fiddle::TYPE_INT
  )
  at_fdcwd = RUBY_PLATFORM.include?("darwin") ? -2 : -100
  Fiddle.last_error = 0
  rc = @linkat.call(at_fdcwd, old_host, at_fdcwd, new_host, 0)
  rc.zero? ? 0 : Fiddle.last_error
end
private :linkat_nofollow
